recipes/mipro/mipro.ipynb

```diff
@@ -0,0 +1,845 @@
+    "    demonstrations = \"\"\n",
+    "    for _, row in bootstrap_sample.iterrows():\n",
+    "        demonstrations += f\"Input: {row[input_col]}\\nOutput: {row[output_col]}\\n\\n\"\n",
+    "\n",
+    "    return demonstrations"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "candidate_demonstrations = [\n",
+    "    generate_demonstrations(\n",
+    "        df=train_df, input_col=\"input_str\", output_col=\"value_str\", seed=seed\n",
+    "    )\n",
+    "    for seed in range(NUM_CANDIDATE_DEMONSTRATIONS)\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Optimize the Prompt\n",
+    "\n",
+    "### Define the optimization objective"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# --- Initialize Online Statistics ---\n",
+    "num_instructions = len(candidate_instructions)\n",
+    "num_demonstrations = len(candidate_demonstrations)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "async def objective(trial: optuna.Trial):\n",
+    "    # Sample an instruction and a demonstration set.\n",
+    "    instruction_index = trial.suggest_categorical(\n",
+    "        \"instruction_index\", range(num_instructions)\n",
+    "    )\n",
+    "    demonstration_index = trial.suggest_categorical(\n",
+    "        \"demonstration_index\", range(num_demonstrations)\n",
+    "    )\n",
+    "    # Asynchronously generate answers for each query in the evaluation set.\n",
+    "    responses = await tqdm_asyncio.gather(\n",
+    "        *[\n",
+    "            generate_answer(\n",
+    "                client=tensorzero_client,\n",
+    "                function_name=\"generate_answer_json\",\n",
+    "                instruction=candidate_instructions[instruction_index],\n",
+    "                demonstrations=candidate_demonstrations[demonstration_index],\n",
+    "                query=query,\n",
+    "                output_schema=user_output_schema,\n",
+    "                system_args=system_args,\n",
+    "                semaphore=semaphore,\n",
+    "            )\n",
+    "            for query, system_args in zip(eval_df[\"input_str\"], eval_df[\"system_args\"])\n",
+    "        ]\n",
+    "    )\n",
+    "\n",
+    "    # Score the responses using the judge.\n",
+    "    judge_responses = await tqdm_asyncio.gather(\n",
+    "        *[\n",
+    "            judge_answer(\n",
+    "                client=tensorzero_client,\n",
+    "                task_description=TASK_DESCRIPTION,\n",
+    "                metric_properties=METRIC_PROPERTIES,\n",
+    "                prediction=str(response.output.parsed) if response is not None else \"\",\n",
```

Comment:
I don't think this works for chat as there's no `parsed` field, needs to be handled differently